### PR TITLE
don't allow empty database names or user names for Postgres connections

### PR DESCRIPTION
**Pull request by capnspacehook**
Fixes https://github.com/gravitational/teleport-private/issues/1507.

changelog: prevent RBAC bypass for new Postgres connections

### CODE REVIEW

File: lib/srv/db/postgres/engine.go [hunk 1]
```diff
@@ -227,6 +227,18 @@ func (e *Engine) handleStartup(client *pgproto3.Backend, sessionCtx *common.Sess
...
+	// the user name as the database name if no database name is passed which
+	// can cause Teleport to allow connections that are explicitly blocked by
+	// RBAC rules. Follow Postgres' behavior and use the user name as the
+	// database name if no database name is passed.
+	if sessionCtx.DatabaseUser == "" {
+		return trace.BadParameter("user name must not be empty")
+	}
+	if sessionCtx.DatabaseName == "" {
+		sessionCtx.DatabaseName = sessionCtx.DatabaseUser
+	}
```
Comment (zmb3) [inline lib/srv/db/postgres/engine.go:240]:
```suggestion
	sessionCtx.DatabaseName = cmp.Or(sessionCtx.DatabaseName, sessionCtx.DatabaseUser)
```

Comment (capnspacehook) [reply lib/srv/db/postgres/engine.go]: Thanks, I always forget about `cmp.Or`

Comment (zmb3) [reply lib/srv/db/postgres/engine.go]: cmp.Or is the real MVP

**Review by greedy52 — APPROVED** [review]

**Review by GavinFrazar — APPROVED** [review]

File: lib/srv/db/postgres/engine.go [hunk 2]
```diff
@@ -227,6 +228,16 @@ func (e *Engine) handleStartup(client *pgproto3.Backend, sessionCtx *common.Sess
...
 func (e *Engine) checkAccess(ctx context.Context, sessionCtx *common.Session) error {
+	// Don't allow an empty user names. Postgres will silently use
+	// the user name as the database name if no database name is passed which
+	// can cause Teleport to allow connections that are explicitly blocked by
+	// RBAC rules. Follow Postgres' behavior and use the user name as the
+	// database name if no database name is passed.
+	if sessionCtx.DatabaseUser == "" {
+		return trace.BadParameter("user name must not be empty")
+	}
+	sessionCtx.DatabaseName = cmp.Or(sessionCtx.DatabaseName, sessionCtx.DatabaseUser)
```
Comment (codingllama) [inline lib/srv/db/postgres/engine.go:239]: TIL cmp.Or.

**Review by codingllama — APPROVED** [review]
LGTM.

**Pull request merged**